java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java

```diff
@@ -0,0 +1,317 @@
+    } else {
+      executionState = blockES.parent.merge(blockES);
+    }
+  }
+
+  @Override
+  public void visitIfStatement(IfStatementTree tree) {
+    scan(tree.condition());
+    ExecutionState thenES = new ExecutionState(executionState);
+    executionState = thenES;
+    scan(tree.thenStatement());
+
+    if (tree.elseStatement() == null) {
+      executionState = thenES.parent.merge(thenES);
+    } else {
+      ExecutionState elseES = new ExecutionState(thenES.parent);
+      executionState = elseES;
+      scan(tree.elseStatement());
+      executionState = thenES.parent.overrideBy(thenES.merge(elseES));
+    }
+  }
+
+  @Override
+  public void visitSwitchStatement(SwitchStatementTree tree) {
+    scan(tree.expression());
+    ExecutionState resultingES = new ExecutionState(executionState);
+    executionState = new ExecutionState(executionState);
+    for (CaseGroupTree caseGroupTree : tree.cases()) {
+      for (StatementTree statement : caseGroupTree.body()) {
+        if (isBreakOrReturnStatement(statement)) {
+          resultingES = executionState.merge(resultingES);
+          executionState = new ExecutionState(resultingES.parent);
+        } else {
+          scan(statement);
+        }
+      }
+    }
+    if (!lastStatementIsBreakOrReturn(tree)) {
+      // merge the last execution state
+      resultingES = executionState.merge(resultingES);
+    }
+
+    if (switchContainsDefaultLabel(tree)) {
+      // the default block guarantees that we will cover all the paths
+      executionState = resultingES.parent.overrideBy(resultingES);
+    } else {
+      executionState = resultingES.parent.merge(resultingES);
+    }
+  }
+
+  private boolean isBreakOrReturnStatement(StatementTree statement) {
+    return statement.is(Tree.Kind.BREAK_STATEMENT, Tree.Kind.RETURN_STATEMENT);
+  }
+
+  private boolean switchContainsDefaultLabel(SwitchStatementTree tree) {
+    for (CaseGroupTree caseGroupTree : tree.cases()) {
+      for (CaseLabelTree label : caseGroupTree.labels()) {
+        if ("default".equals(label.caseOrDefaultKeyword().text())) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private boolean lastStatementIsBreakOrReturn(SwitchStatementTree tree) {
+    List<CaseGroupTree> cases = tree.cases();
+    if (!cases.isEmpty()) {
+      List<StatementTree> lastStatements = cases.get(cases.size() - 1).body();
+      return !lastStatements.isEmpty() && isBreakOrReturnStatement(lastStatements.get(lastStatements.size() - 1));
+    }
+    return false;
+  }
+
+  @Override
+  public void visitWhileStatement(WhileStatementTree tree) {
+    scan(tree.condition());
+    visitStatement(tree.statement());
+  }
+
+  @Override
+  public void visitDoWhileStatement(DoWhileStatementTree tree) {
+    visitStatement(tree.statement());
+    scan(tree.condition());
+  }
+
+  @Override
+  public void visitForStatement(ForStatementTree tree) {
```

Comment:
from an execution point of view visit should be done in that order:
initializer, condition, update, statement